feat: use Segment (batch 1)

NicolasMassart opened the pull request:
## **Description**

Batch PR to group changes from:
- #7520
- #7990
- #8059 
- #8090
- #8218
- #8471
- #8573 

## **Related issues**

Fixes https://github.com/MetaMask/mobile-planning/issues/1129

## **Manual testing steps**

See each PR for details

## **Screenshots/Recordings**

See each PR for details

### **Before**

See each PR for details

### **After**

See each PR for details

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
  - [x] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

NicolasMassart committed:
feat: create Segment analytics implementation + use geo-localization proxy (Batch 2 of 4) (#7520)

# What is the reason for the change?
- Implement a new metrics system for the Segment SDK.
- Make it work alongside the old one so that we can remove the old one
progressively.

# What is the improvement/solution?
- Create a new class to handle metrics with segment
- abstract all segment SDK calls
- provide our own user and anonymous Ids for improved privacy
- Reuse the Mixpanel user id it available, otherwise generates a new one
with the same format
- patch Segment SDK to allow the privacy aware user and anonymous Ids
- No event sent at all if user haven't accepted to send metrics info
- Total reset of the infos in case user asks to reset (in settings)
- Ability to disable all metrics
- Ability to ask for deletion of all data

Fixes MetaMask/mobile-planning#1222

NicolasMassart committed:
Merge branch 'main' into feat/batch_1129_segment

NicolasMassart committed:
Merge branch 'main' into feat/batch_1129_segment

NicolasMassart committed:
Merge branch 'main' into feat/batch_1129_segment

NicolasMassart committed:
feat: Use new Segment implementation in onboarding steps (Batch 3 of 4) (#7990)

- Initialise MetaMetrics on app start without removing Analytics yet as
we still uses it outside of onboarding
- create a `trackAfterInteractions` utility function to reduce code when
tracking events.
- Replace calls to Analytics with calls to `trackAfterInteractions` in
all components involved in onboarding screens (see
[`OnboardingNav`](https://github.com/MetaMask/metamask-mobile/blob/feat%2F1275_segment_onboarding_impl/app/components/Nav/App/index.js#L122-L175)
in `app/components/Nav/App/index.js`)
- Fix onboarding events queue usage in
`app/components/UI/OptinMetrics/index.js`
  - Update Opt-on/Opt-out in onboarding
- Extract Analytics User profil properties in UserProfileMetaData util
and inject it in MetaMetrics identify request alongside device metadata

Fixes https://github.com/MetaMask/mobile-planning/issues/1275

NicolasMassart committed:
fix: rename Segment env vars and add in example file (#8059)

- rename Segment env vars:
  - `SEGMENT_WRITE_KEY`
  - `SEGMENT_PROXY_URL`
- add env var placeholders in example file

Fixes missing variables naming change in
https://github.com/MetaMask/mobile-planning/issues/1222

NicolasMassart committed:
Merge branch 'main' into feat/batch_1129_segment

NicolasMassart committed:
Merge branch 'feat/batch_1129_segment' of github.com:MetaMask/metamask-mobile into feat/batch_1129_segment

NicolasMassart committed:
Merge branch 'main' into feat/batch_1129_segment

NicolasMassart committed:
Merge branch 'main' into feat/batch_1129_segment

NicolasMassart committed:
Merge branch 'main' into feat/batch_1129_segment

# Conflicts:
#	app/components/Views/AccountBackupStep1/index.js
#	app/components/Views/AccountBackupStep1B/index.js
#	app/components/Views/ChoosePassword/index.js
#	app/components/Views/ManualBackupStep1/index.js
#	app/components/Views/ManualBackupStep2/index.js
#	app/components/Views/ManualBackupStep3/index.js

NicolasMassart committed:
Merge branch 'main' into feat/batch_1129_segment

NicolasMassart committed:
Merge branch 'main' into feat/batch_1129_segment

# Conflicts:
#	app/components/Views/Settings/SecuritySettings/SecuritySettings.tsx

NicolasMassart committed:
Merge branch 'main' into feat/batch_1129_segment

NicolasMassart committed:
Merge branch 'main' into feat/batch_1129_segment

NicolasMassart committed:
Merge branch 'main' into feat/batch_1129_segment

NicolasMassart committed:
Merge branch 'main' into feat/batch_1129_segment

# Conflicts:
#	yarn.lock

NicolasMassart committed:
Merge branch 'main' into feat/batch_1129_segment

NicolasMassart committed:
Merge branch 'main' into feat/batch_1129_segment

NicolasMassart committed:
feat: segment data deletion (4/4) (#8090)

- adds `createDataDeletionTask` to `MetaMetrics` to call the deletion
API in MetaMetrics class for method [create source
regulation](https://docs.segmentapis.com/tag/Deletion-and-Suppression#operation/createSourceRegulation)
- adds the `checkDataDeletionTaskStatus` to `MetaMetrics` to call
deletion API [get regulation
method](https://docs.segmentapis.com/tag/Deletion-and-Suppression#operation/getRegulation)
- calls the `createDataDeletionTask` and `checkDataDeletionTaskStatus`
from `DeleteMetaMetricsData.tsx` when user touches the delete button.
- extract the settings section logic to display allow deletion to a well
tested hook.
- Remove `Analytics.js` implementation of these features and point only
to the new `MetaMetrics` ones.
- update unit tests
- rename in `storage.js`:
- `METAMETRICS_SEGMENT_REGULATION_ID` ->
`METAMETRICS_DELETION_REGULATION_ID`

Fixes https://github.com/MetaMask/mobile-planning/issues/1403

NicolasMassart committed:
Merge branch 'main' into feat/batch_1129_segment

# Conflicts:
#	ios/Podfile.lock

NicolasMassart committed:
feat: Segment instance error handling (#8218)

-adds error handling on async `DefaultPreference.get` and
`DefaultPreference.set`
- makes building instance stronger by using the Config type to prevent
config mistakes

completes https://github.com/MetaMask/mobile-planning/issues/1403

NicolasMassart committed:
Merge branch 'main' into feat/batch_1129_segment

worldlyjohn commented on app/core/Analytics/MetaMetrics.ts:
```diff
@@ -1,314 +1,654 @@
+    if (!this.instance) {
+      const config: Config = {
+        writeKey: process.env.SEGMENT_WRITE_KEY as string,
+        proxy: process.env.SEGMENT_PROXY_URL as string,
```
are we planning to use a separate endpoint for PROD from the rest? (cc @vu1n)

NicolasMassart replied:
Yes if we have one provided. The endpoint is provided by env variable so it can be different depending on the build type.

NicolasMassart committed:
Merge branch 'main' into feat/batch_1129_segment

# Conflicts:
#	yarn.lock

NicolasMassart committed:
update lockfile

NicolasMassart committed:
Merge branch 'main' into feat/batch_1129_segment

NicolasMassart committed:
feat: a hook to use Segment metrics (#8471)

- make MetaMetrics.getInstance synchronous and add an async init
function + readiness check
- destructurable object hook return value to use metrics in components
without having to use useRef and useEffect to set the instance
- interface for the destructurable object
- HOC (High Order Component) wrapper for class components
- new and updated unit tests

NicolasMassart committed:
Merge branch 'main' into feat/batch_1129_segment

NicolasMassart committed:
Merge branch 'main' into feat/batch_1129_segment

NicolasMassart committed:
feat: migrate Segment core mechanism [split of 8387] (#8573)

update metrics system for missing features when migrating all app
- add getMetaMetricsId
- add exports
- fix/update tests
- fix HOC types

NicolasMassart committed:
Merge branch 'main' into feat/batch_1129_segment

NicolasMassart committed:
Merge branch 'main' into feat/batch_1129_segment

NicolasMassart committed:
fix import and getInstance call

NicolasMassart committed:
remove await

NicolasMassart committed:
use hook

NicolasMassart committed:
Merge branch 'main' into feat/batch_1129_segment

sethkfman commented on app/core/Analytics/MetaMetrics.constants.ts:
```diff
@@ -1,4 +1,4 @@
 
 export const SEGMENT_REGULATIONS_ENDPOINT =
-  'https://api.segmentapis.com/regulations';
+  'https://proxy.metafi-dev.codefi.network/segment/v1';
```
This should be updated to the prod URL.

NicolasMassart committed:
use env var for endpoint

NicolasMassart committed:
add test env for metametrics

NicolasMassart committed:
make type for track consistent between class and hook

tommasini commented on app/components/Views/Settings/SecuritySettings/Sections/DeleteMetaMetricsData.tsx:
```diff
@@ -115,7 +167,7 @@ const DeleteMetaMetricsData = () => {
       sectionButtonText={strings('app_settings.delete_metrics_button')}
       descriptionText={
-        hasCollectedData ? (
+        isDataDeletionAvailable() ? (
```
We could call this function only once and have the variable being used here

tommasini commented on app/components/Views/Settings/SecuritySettings/Sections/DeleteMetaMetricsData.tsx:
```diff
@@ -156,7 +208,7 @@ const DeleteMetaMetricsData = () => {
         )
       }
-      buttonDisabled={!hasCollectedData}
+      buttonDisabled={!isDataDeletionAvailable()}
```
And here

tommasini reviewed (approved):
AMAZING WORK! LGTM! Just left a nit pick that i've found but it's okay to merge it without it!

tommasini reviewed (changes requested):
Actually it seems that the unit tests are failing! We will need to address that

NicolasMassart committed:
fix configure

remove useless async
move log to configure
fix trackEvent param type
update test

NicolasMassart committed:
fix test

NicolasMassart committed:
fix type

tommasini commented on app/components/Nav/App/index.js:
```diff
@@ -370,15 +369,14 @@ const App = ({ userLoggedIn }) => {
     const initAnalytics = async () => {
-      const metaMetricsId = await getMetaMetricsId();
-      Logger.log(`MetaMetrics initialised with ID: ${metaMetricsId}`);
+      await MetaMetrics.getInstance().configure();
```
No more Logger log with the instance id?

tommasini commented on app/core/Analytics/MetaMetrics.ts:
```diff
@@ -507,6 +505,8 @@ class MetaMetrics implements IMetaMetrics {
       this.dataRecorded = await this.#getIsDataRecordedFromPrefs();
       this.#isConfigured = true;
+      if (__DEV__)
+        Logger.log(`MetaMetrics configured with ID: ${this.metametricsId}`);
```
Cool!

tommasini reviewed (dismissed):
LGTM!

NicolasMassart committed:
optimise metrics settings

tommasini reviewed (approved):
LGTM!
